apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ReliableStreamReader.cs

```diff
@@ -0,0 +1,258 @@
+using System.Threading.Tasks;
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Cloud.Spanner.V1.Logging;
+using Google.Protobuf;
+using Google.Protobuf.WellKnownTypes;
+using Grpc.Core;
+
+// ReSharper disable PrivateFieldCanBeConvertedToLocalVariable
+// ReSharper disable NotAccessedField.Local
+// ReSharper disable EmptyDestructor
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// Provides streaming access to a Spanner SQL query that automatically retries, handles
+    /// chunking and recoverable errors.
+    /// </summary>
+    public sealed class ReliableStreamReader : IDisposable
+    {
+        private AsyncServerStreamingCall<PartialResultSet> _currentCall;
+        private readonly object _sync = new object();
+        private readonly SpannerClient _spannerClient;
+        private readonly ExecuteSqlRequest _request;
+        private readonly Session _session;
+        private readonly CallSettings _callTiming;
+        private ResultSetMetadata _metadata;
+        private bool _isReading = true;
+
+        private ByteString _resumeToken;
+        private int _resumeSkipCount;
+
+        internal ReliableStreamReader(SpannerClient spannerClientImpl, ExecuteSqlRequest request, Session session)
+        {
+            _spannerClient = spannerClientImpl;
+            _request = request;
+            _session = session;
+            _clock = SpannerSettings.GetDefault().Clock ?? SystemClock.Instance;
+            _scheduler = SpannerSettings.GetDefault().Scheduler ?? SystemScheduler.Instance;
+
+            _callTiming = CallSettings.FromCallTiming(CallTiming.FromRetry(new RetrySettings(
+                    SpannerSettings.GetDefaultRetryBackoff(),
+                    SpannerSettings.GetDefaultTimeoutBackoff(),
+                    Expiration.FromTimeout(TimeSpan.FromMilliseconds(600000)),
+                    SpannerSettings.IdempotentRetryFilter
+                )));
+            _request.SessionAsSessionName = _session.SessionName;
+        }
+
+        private async Task<Metadata> ConnectAsync()
+        {
+            Logger.LogPerformanceCounterFn("StreamReader.ConnectCount", x => x + 1);
+            if (_resumeToken != null)
+            {
+                Logger.Debug(() => $"Resuming at location:{_resumeToken}");
+                _request.ResumeToken = _resumeToken;
+            }
+            _currentCall = _spannerClient.ExecuteSqlStream(_request);
+            return await _currentCall.ResponseHeadersAsync.ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Connects or reconnects to Spanner, fast forwarding to where we left off based on
+        /// our _resumeToken and _resumeSkipCount.
+        /// </summary>
+        /// <returns></returns>
+        private async Task<bool> ReliableConnect(CancellationToken cancellationToken)
+        {
+            if (_currentCall == null)
+            {
+//                Func<Task<Metadata>> connectFn = ConnectAsync;
```

Comment:
removed for now
this is using my extension method for doing retries -- which is my fallback plan in case chris bacon doesn't provide the retry in the vkit layer.  Once he provides that work, I'll add back the calltiming config above and pass it into him to set up retries.

So right now, ReliableStreamReader does not retry when doing the initial connect, but will recover when the stream is broken by reconnecting with the correct resume token.